posix: separate directory for shell utilities and menuconfig cleanup

cfriedt opened the pull request:
* separate shell utilities ("shell") from POSIX API implementation ("options") with directories
* use `menu` and `menuconfig` Kconfig directives where appropriate / possible to clean up menuconfig
* remove unnecessary `APP_LINK_WITH_POSIX_SUBSYSTEM`, since private headers should be private and there is no need for the application to include private posix headers
* add minor path changes to `samples/posix/philosophers`
* fix typos and other rough edges in `samples/posix/philosophers` and `samples/philosophers`

Before
<img width="490" alt="Screenshot 2024-01-27 at 12 46 20 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/bb37fd49-1f60-4c9d-80d4-ec7208c87b5e">

After
<img width="499" alt="Screenshot 2024-01-29 at 8 33 42 AM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/004fccc5-08de-45b5-bb76-ea775b7347d3">
<img width="482" alt="Screenshot 2024-01-31 at 3 44 53 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/e09569da-5eeb-47a1-96b8-8e1131157869">

cfriedt commented:
Going out on a limb here, but I'm going to guess that this error is quite unrelated.
```
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_tiny@768/samples/philosophers/sample.kernel.philosopher.same_prio/handler.log
INFO    - 701/704 qemu_x86_tiny@768         samples/philosophers/sample.kernel.philosopher.fifos  FAILED Timeout (qemu 59.957s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_tiny@768/samples/philosophers/sample.kernel.philosopher.fifos/handler.log
ERROR   - SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..*** Booting Zephyr OS build zephyr-v3.5.0-4[941](https://github.com/zephyrproject-rtos/zephyr/actions/runs/7702666462/job/20991599646?pr=68188#step:13:942)-g0f72df763786 ***
ASSERTION FAIL [!arch_is_in_isr()] @ WEST_TOPDIR/zephyr/kernel/include/ksched.h:255
```

ycsin commented:
🤔

```
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_tiny@768/samples/philosophers/sample.kernel.philosopher.lifos/handler.log
INFO    - 699/704 qemu_x86_tiny@768         samples/philosophers/sample.kernel.philosopher.fifos  FAILED Timeout (qemu 59.967s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_tiny@768/samples/philosophers/sample.kernel.philosopher.fifos/handler.log
ERROR   - SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..*** Booting Zephyr OS build zephyr-v3.5.0-4943-gcde3d755e718 ***
ASSERTION FAIL [!arch_is_in_isr()] @ WEST_TOPDIR/zephyr/kernel/include/ksched.h:255
```

cfriedt commented:
> 🤔

I know... it seems to be unrelated, right?

ycsin reviewed (commented):
The assertion error in CI also occurs in `main`

cfriedt commented on samples/posix/philosophers/README.rst:
```diff
@@ -13,8 +13,7 @@ This sample implements Zephyr's :ref:`Dining Philosophers Sample <dining-philoso
 
-This project outputs to the console.  It can be built and executed
-on QEMU as follows:
+This project outputs to the console.  It can be built and executed on QEMU as follows:
```
If it's ok, I'll wait for another round of reviews before fixing.

kartben replied:
It's fine if you don't bother, I guess :)

<img width="144" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/f0f80f8f-90e2-4b4c-892e-da53883a6e72">

cfriedt replied:
Fixed!

jukkar reviewed (dismissed):
LGTM

kartben reviewed (dismissed):
+1 for the sample fixes. Please check my non-blocking comment if you have a chance

kartben commented on samples/posix/philosophers/prj.conf:
This looks suspiciously crowded for what I would expect to be a reasonably simple sample - any chance you could quickly check there's nothing that's redundant / not strictly necessary (esp. since some options were carried over from the original sample)? It's always nice to try to keep samples as minimal as can be.

cfriedt replied:
I'll see if there are any that are redundant. Chances are, they might apply to the original as well.

cfriedt replied:
Fixed!

kartben replied:
Neat!

Commit:
posix: separate shell utilities and posix api implementation

Previously, the POSIX shell utilities were intermixed with the
POSIX API implementation.

The POSIX shell utilities only depend on the public POSIX API,
so it makes sense to keep them in a separate subdirectory.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: remove unneeded option to link with posix subsys

The CONFIG_APP_LINK_WITH_POSIX_SUBSYS option was originally
present so that internal POSIX implementation headers would be
on the include path.

There is no implicit need for any app or library to include
private POSIX headers. Instead, the standard POSIX API should
be used.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
samples: philosophers correct misspelled introduction tag

The tag 'introduction' was misspelled 'inroduction'.

Correct it both the samples/philosophers and the
samples/posix/philosophers directories.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

kartben commented:
@cfriedt looks like you pushed the wrong branch

Commit:
samples: posix: philosophers: minor spacing corrections

Fix spacing and wrapping in the README.rst file.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
samples: philosophers: remove redundant items in prj.conf

The normal and posix philosophers samples seemed to have
a number of redundant choices in prj.conf.

Let's reduce to the minimum required for the sample. The
CONFIG_DEBUG_THREAD_INFO option is part of the
documentation, so leave it as-is.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
samples: posix: philosophers: fix off-by-one error

The number of forks should match the number of philophers, and
that is encoded via NUM_PHIL.

Change the build assert to match.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
> @cfriedt looks like you pushed the wrong branch

Yes - it took a while, but it should be fixed.. have to let it run through CI and get some 💤

ycsin reviewed (dismissed):
LGTM, thanks for this cleanup!

CI might fail due to #68272 though, because the modified POSIX philosophers sample share the same tags as the normal philosophers, but that should be unrelated.

cfriedt commented:
@jukkar - is it possible to get your +1 again?

kartben reviewed (dismissed):
Changes related to Philosophers sample look good, thx! 
I didn't review the other changes.

cfriedt commented:
Will rebase again to fix merge conflicts

cfriedt commented:
@ycsin @jukkar @kartben - sorry to both again, but please re-approve. Ideally this would get in before others in the POSIX so that there are no more merge conflicts.
